services/streammanager/googlesheets/googlesheetsmanager.go

```diff
@@ -0,0 +1,236 @@
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/tidwall/gjson"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/sheets/v4"
+)
+
+type Config struct {
+	Credentials     string              `json:"credentials"`
+	EventToSheetMap []map[string]string `json:"eventToSpreadSheetIdMap"`
+}
+
+type Credentials struct {
+	Email      string `json:"client_email"`
+	PrivateKey string `json:"private_key"`
+	TokenUrl   string `json:"token_uri"`
+}
+
+type GoogleAPIService struct {
+	Jwt     *jwt.Config
+	Service *sheets.Service
+}
+
+var pkgLogger logger.LoggerI
+var googleAPIService GoogleAPIService
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child("googlesheets")
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destinationConfig interface{}) (*GoogleAPIService, error) {
+	var config Config
+	var credentialsFile Credentials
+	jsonConfig, err := json.Marshal(destinationConfig)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleSheets] Error while marshalling destination config :: %w", err)
+	}
+	err = json.Unmarshal(jsonConfig, &config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleSheets] error  :: error in GoogleSheets while unmarshalling destination config:: %w", err)
+	}
+	if config.Credentials != "" {
+		err = json.Unmarshal([]byte(config.Credentials), &credentialsFile)
+		if err != nil {
+			return nil, fmt.Errorf("[GoogleSheets] error  :: error in GoogleSheets while unmarshalling credentials json:: %w", err)
+		}
+
+	}
+	// Creating token URL from Credentials file if not using constant from google.JWTTOkenURL
+	tokenURI := google.JWTTokenURL
+	if len(credentialsFile.TokenUrl) != 0 {
+		tokenURI = credentialsFile.TokenUrl
+	}
+	// Creating JWT Config which we are using for getting the oauth token
+	jwtconfig := &jwt.Config{
+		Email:      credentialsFile.Email,
+		PrivateKey: []byte(credentialsFile.PrivateKey),
+		Scopes: []string{
+			"https://www.googleapis.com/auth/spreadsheets",
+		},
+		TokenURL: tokenURI,
+	}
+	googleAPIService.Jwt = jwtconfig
+	service, err := GenerateServiceWithrefreshToken(*jwtconfig)
+	googleAPIService.Service = service
+	return &googleAPIService, err
+}
+
+func Produce(jsonData json.RawMessage, producer interface{}, destConfig interface{}) (statusCode int, respStatus string, responseMessage string) {
+
+	// Here we recieve the transformed json with all the required mappings
+	parsedJSON := gjson.ParseBytes(jsonData)
+	// Storing sheet_id and sheet_tab for the specific message
+	spreadSheetId := parsedJSON.Get("spreadSheetId").String()
+	spreadSheetTab := parsedJSON.Get("spreadSheetTab").String()
+
+	var pServiceMessage string
+	// First we are pinging to google-sheets using existing client
+	pkgLogger.Debug("[Google Sheets] Pingging to Sheets using existing client", googleAPIService)
```

Comment:
Do we need to ping first? We will end up doing 2 calls for every event. 

Instead, can we try to post and if it fails , then based on the error code - refresh the token.

Reply:
Did a revamp of the existing code to support this. Added a `func` to handle insertion of data using sheets-service, which internally handles if there is a token issue in the error by creating a new client and calls itself recursively in order to prevent the current event-loss.

Add inline comments explaining functionality